Tiny obj loader

nkoenig opened the pull request:
This pull requests includes tinyobjloader, an external OBJ loader.

`Mesh::LoadObjFile` is updated to use tinyobjloader.

I have tested this with `ctest -VV -R model_test`

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3805)

<!-- Reviewable:end -->

sherm1 commented:
Some minor build failures detected in CI: https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=148134

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KU3byUcwnwqYsmT77eF:b-s2pd0m)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@nkoenig, TRI convention is to have the PR author propose a feature reviewer and a platform reviewer (usually the on-call [platform-reviewer-of-the-day](https://sites.google.com/a/tri.global/internal/drake-platform-review)). For feature review I would suggest Liang +@liangfok, and today's platform reviewer is Grant +@ggould-tri.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KU3cYSCCjPSiPL8OHhB:b-cw5a8p)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
@nkoenig Thanks for this!  A few nits below for conforming to our style guide; <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once those resolve.

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 273 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eaCWwiusPxBiuw9m:-KU3eaCWwiusPxBiuw9n:bqk2b2c) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L273)):*
> ```cpp
>   size_t idx = obj_file_name.rfind('/');
>   if (idx != string::npos)
>     path = obj_file_name.substr(0, idx+1);
> ```

Code conventions require braces `{}` around multiline if statements.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 283 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3elIoktcKpdWCU_GL:-KU3elIpOY7EGWKkkpXa:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L283)):*
> ```cpp
>       obj_file_name.c_str(), path.c_str(), true);
> 
>   // Report any error
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 287 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3er87DqkmtAStEMBt:-KU3er88GHjqmoVRVXRC:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L287)):*
> ```cpp
>     std::cerr << "Error loading OBJ file: " << err << std::endl;
> 
>   // Use the boolean return value to determine if we should proceeed
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 291 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eraPK6hCWRmdh0i9:-KU3eraPK6hCWRmdh0iA:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L291)):*
> ```cpp
>     throw std::runtime_error("Error parsing file \"" + obj_file_name + "\".");
> 
>   // Store the vertices
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 298 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3et4B8E8V82WVAJnY:-KU3et4B8E8V82WVAJnZ:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L298)):*
> ```cpp
>   }
> 
>   // Iterate over the shapes
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 305 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3fil1NGpEPGYzU6UX:-KU3fil2giN-1R8_HfcU:b-8x793c) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L305)):*
> ```cpp
>     for (unsigned int face = 0;
>          face < shape.mesh.num_face_vertices.size(); ++face) {
>       unsigned int vertCount = shape.mesh.num_face_vertices[face];
> ```

Code conventions https://google.github.io/styleguide/cppguide.html#Integer_Types discourage use of named integer types and http://drake.mit.edu/code_style_guide.html#clarifications forbid unsigned types except for `size_t`.  I suggest `size_t` here.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eu_OlHTD0YTp3XVX:-KU3eu_OlHTD0YTp3XVY:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L307)):*
> ```cpp
>       unsigned int vertCount = shape.mesh.num_face_vertices[face];
> 
>       // Make sure the face has three vertices
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 319 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3gVNZVYjHCKMb_6Tm:-KU3gVNZVYjHCKMb_6Tn:bj2wj1n) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L319)):*
> ```cpp
>       // For each vertex in the face
>       for (unsigned int vert = 0; vert < vertCount; ++vert) {
>         // Store the vertex index
> ```

FYI code conventions require a period at the end of complete sentences in comments.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KU3gihhuewLavQuvZ2D:blvpc3z)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
@nkoenig, you might find `clang_format` useful for making your code conform. Drake-specific instructions [here](http://drake.mit.edu/code_style_tools.html). Unfortunately it's too feeble to fix commenting style, but it knows where to put curlies, indents, and the like.

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KU3umn-P9zyAQiIVGTn:bkh6siy)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I completed my first round of review. Looks good overall. Just a few minor comments.

---

Reviewed 5 of 6 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 285 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3uJ8PTVqQfZc6EZrZ:-KU3uJ8PTVqQfZc6EZr_:b6xgmak) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L285)):*
> ```cpp
>   // Report any error
>   if (!err.empty())
>     std::cerr << "Error loading OBJ file: " << err << std::endl;
> ```

BTW: This error message should ideally be moved into the exception that's thrown below.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 292 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3w6VaQ-au5NAzwl1q:-KU3w6VaQ-au5NAzwl1r:b-h9myny) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L292)):*
> ```cpp
> 
>   // Store the vertices
>   for (size_t idx = 0; idx < attrib.vertices.size(); idx += 3) {
> ```

Our style guide disallows the use of `size_t` here.
http://drake.mit.edu/code_style_guide.html#clarifications

---

*[drake/systems/plants/shapes/Geometry.cpp, line 299 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3tONv3b3YSG4abgTq:-KU3tONwnyj2ZUVVO4dJ:b3gg0pt) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L299)):*
> ```cpp
> 
>   // Iterate over the shapes
>   for (auto const shape : shapes) {
> ```

BTW: Would `auto const&` work here to avoid a copy?

---

*[drake/systems/plants/shapes/Geometry.cpp, line 303 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3wTS_KKcyCIwG65E_:-KU3wTS_KKcyCIwG65Ea:b-r3uieo) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L303)):*
> ```cpp
> 
>     // For each face in the shape
>     for (unsigned int face = 0;
> ```

Our style guide disallows the use of `unsigned int` here.

http://drake.mit.edu/code_style_guide.html#clarifications

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KU3wj52LX9DZZ9ECzy2:b-uxxnli)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 292 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3w6VaQ-au5NAzwl1q:-KUHsTLUG7XS-6UqT6Gl:b-wkr6ij) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L292)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Our style guide disallows the use of `size_t` here.
> http://drake.mit.edu/code_style_guide.html#clarifications
</details>
@liangfok `std::vector::size` is exactly the case envisioned by "a (1) non-Drake API uses unsigned types" in the guide.  #2514 specifically calls out `std::vector` as one of the cases where we are obliged to permit `size_t`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 292 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3w6VaQ-au5NAzwl1q:-KUI3K-rf8fvy2Kk_8YE:b-1ynp22) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L292)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> @liangfok `std::vector::size` is exactly the case envisioned by "a (1) non-Drake API uses unsigned types" in the guide.  #2514 specifically calls out `std::vector` as one of the cases where we are obliged to permit `size_t`.
</details>
I see. I guess I was confused b/c there are certainly examples of the return value of `std::vector::size()` being cast into an `int` in Drake. For example, see:

https://github.com/RobotLocomotion/drake/blob/b0f3ab4a6dc9c1670cea9354744a6dd43cf8b602/drake/systems/framework/diagram.h#L354

I'll mark this thread satisfied and leave it to the author's discretion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 5 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 273 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eaCWwiusPxBiuw9m:-KUNLV3tYiA1zdyAMxz3:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L273)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Code conventions require braces `{}` around multiline if statements.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 283 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3elIoktcKpdWCU_GL:-KUNLWfRZxqt7eaj1ldV:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L283)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 285 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3uJ8PTVqQfZc6EZrZ:-KUNLY-EpAD96Yh-bBXZ:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L285)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: This error message should ideally be moved into the exception that's thrown below.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 287 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3er87DqkmtAStEMBt:-KUNLYxV6whzmKH0EeSP:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L287)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 291 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eraPK6hCWRmdh0i9:-KUNL_0UniPHuIVU4Bxp:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L291)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 292 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3w6VaQ-au5NAzwl1q:-KUNLdyoETZMueg2xnXG:brvricv) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L292)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I see. I guess I was confused b/c there are certainly examples of the return value of `std::vector::size()` being cast into an `int` in Drake. For example, see:
> 
> https://github.com/RobotLocomotion/drake/blob/b0f3ab4a6dc9c1670cea9354744a6dd43cf8b602/drake/systems/framework/diagram.h#L354
> 
> I'll mark this thread satisfied and leave it to the author's discretion.
</details>
I left this as size_t.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 298 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3et4B8E8V82WVAJnY:-KUNLgcQ-rXJWOpFeX6l:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L298)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 299 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3tONv3b3YSG4abgTq:-KUNLiDrX9rbYhl6FGxi:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L299)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: Would `auto const&` work here to avoid a copy?
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 303 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3wTS_KKcyCIwG65E_:-KUNLjJ6DZlR-3lxFQlR:b6awsy2) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L303)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Our style guide disallows the use of `unsigned int` here.
> 
> http://drake.mit.edu/code_style_guide.html#clarifications
</details>
Done in d82fa3f, changed to auto

---

*[drake/systems/plants/shapes/Geometry.cpp, line 305 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3fil1NGpEPGYzU6UX:-KUNLkaWsZe3ZuQ6tYqr:b-hp38kc) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L305)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> Code conventions https://google.github.io/styleguide/cppguide.html#Integer_Types discourage use of named integer types and http://drake.mit.edu/code_style_guide.html#clarifications forbid unsigned types except for `size_t`.  I suggest `size_t` here.
</details>
Done in d82fa3f, changed to auto.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 307 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3eu_OlHTD0YTp3XVX:-KUNLqCeTrS2Ds7gMFm3:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L307)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---

*[drake/systems/plants/shapes/Geometry.cpp, line 319 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KU3gVNZVYjHCKMb_6Tm:-KUNLs4kfmR25wrtym3W:bhd34p6) ([raw file](https://github.com/robotlocomotion/drake/blob/471d1c4c59a52f158d1b7f429e4c707ceddce27e/drake/systems/plants/shapes/Geometry.cpp#L319)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI code conventions require a period at the end of complete sentences in comments.
</details>
Done in d82fa3f

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Thanks Sherm. I'll make sure to propose a feature reviewer and platform reviewer. I'll propose two at the start of day tomorrow.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KUP82FD0kskgzJCePos:bo52o2m)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
The syntax I used above (plus-at-githubid) is Reviewable magic that assigns reviewers -- I assigned Liang for feature and Grant for platform. However, Grant is off-duty until Friday so when you're ready you can unassign him (with minus-at-ggould-tri) and assign that day's platform reviewer instead. You can of course assign additional reviewers if you want.

---

Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KUPEehhLeDXm6dzf7Ga:b-7ubsgs)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 273 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KUPFxO_BfOlHvB26RGD:-KUPFxO_BfOlHvB26RGE:bm1w03s) ([raw file](https://github.com/robotlocomotion/drake/blob/d82fa3f3e72d88162cf2b989239591b9920d6c97/drake/systems/plants/shapes/Geometry.cpp#L273)):*
> ```cpp
>   size_t idx = obj_file_name.rfind('/');
>   if (idx != string::npos) {
>     path = obj_file_name.substr(0, idx+1);
> ```

BTW: nit: missing spaces around "+".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KUPG5Kwi93UQbfbhaZH:bnfp4nl)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Ahh. I was hoping there was some magic in the pipeline to notify reviewers.

---

Review status: 5 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 273 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KUPFxO_BfOlHvB26RGD:-KUSKtB-mrSkcDKo9va5:buzr2sb) ([raw file](https://github.com/robotlocomotion/drake/blob/d82fa3f3e72d88162cf2b989239591b9920d6c97/drake/systems/plants/shapes/Geometry.cpp#L273)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW: nit: missing spaces around "+".
</details>
Fixed in ba9f210

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KUSKL3urt0PpVsXkots:b5k3q13)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KUSXgzZn-K1sA-xaEqe:-KUSXgzZn-K1sA-xaEqf:bz2q8bw) ([raw file](https://github.com/robotlocomotion/drake/blob/ba9f2102d8fdb9d6989236c236cde413d7b6d909/drake/systems/plants/shapes/Geometry.cpp#L299)):*
> ```cpp
> 
>   // Iterate over the shapes.
>   for (auto const &shape : shapes) {
> ```

The `&` goes next to the type rather than the variable. In other words, this should be:

```
for (auto const& shape : shapes) {
```

Sorry this was unclear, see:

https://reviewable.io/reviews/robotlocomotion/drake/3830#-KUSH9E-rBwyUCb560uX

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
This is <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending resolution of outstanding discussion and build issues.

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KUrJgV13RbA95Nqb3bH:bp2cy2o)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/shapes/Geometry.cpp, line 299 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KUSXgzZn-K1sA-xaEqe:-KV0_4Z1nvzqRDCCQ8g8:b-w5u1mq) ([raw file](https://github.com/robotlocomotion/drake/blob/ba9f2102d8fdb9d6989236c236cde413d7b6d909/drake/systems/plants/shapes/Geometry.cpp#L299)):*
> 01c6dcd:
Changed in 01c6dcd:

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
There is one "variable shadow" build error. Also, this PR needs to be rebased with master.

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KV0l9kwoA6i9hiKHhUz:brv0e9p)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Here is the shadow error: https://drake-cdash.csail.mit.edu/viewBuildError.php?buildid=158577

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KV0udx30szxUtZhKV7c:b6wywhe)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please

jamiesnape commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

jamiesnape commented:
@drake-jenkins-bot retest this please
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-debug please
@drake-jenkins-bot linux-clang-ninja-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please
@drake-jenkins-bot linux-gcc-ninja-experimental-open-source-debug please

ggould-tri commented:
Is anyone still working this PR?  It is lgtm'd but failing CI.
If it is no longer active, I recommend that it be closed.  If it is active, I believe it can be merged, fixed, and landed relatively easily.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KVeNOkLsg0C8Y_M-zzC:b1z2bt4)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Looks like this PR needs to be rebased with master.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KW3U-4xy55ipbef9uIj:b-ndlcan)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@nkoenig This PR needs some attention before it can merge.  What's your thoughts on getting this building again / ready to finish review / merged?

nkoenig commented:
I'm back on it. sorry for the delay.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KXl3XNSPsMlNRm6m-t3:b-v8dtr9)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Left a couple minor comments / questions.

---

Reviewed 5 of 7 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KXr2bGXJVKmyBiAa_YH:baz8sxh) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
> ```cpp
> #include "drake/multibody/shapes/geometry.h"
> #include "externals/tinyobjloader/tiny_obj_loader.h"
> ```

BTW, is "externals/" necessary?

---

*[drake/multibody/shapes/geometry.cc, line 292 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2iPXg3DiU9YgNRD8:-KXr2iPY2bAyaVTGiiSV:bldms2c) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L292)):*
> ```cpp
>   }
> 
>   // Store the vertices
> ```

Missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KXr2nbNktQYHpk0yvJo:b-3ngd6i)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 5 of 6 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KXrDw95GVUykmgqtPnv:b-sbo674) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, is "externals/" necessary?
</blockquote></details>

I see the following error if I remove `externals`:

```
drake/drake/multibody/shapes/geometry.cc:2:43: fatal error: tinyobjloader/tiny_obj_loader.h: No such file or directory
```

Maybe I need to change the `include_directories` in CMakeLists.txt. Checking...

---

*[drake/multibody/shapes/geometry.cc, line 292 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2iPXg3DiU9YgNRD8:-KXrENqWXBHRYfrMQ9iJ:bt2vr1b) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L292)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Missing period at end of sentence.
</blockquote></details>

Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KXrKWdPnEqyDby7UJwD:b-7ubzin) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary><blockquote>

I see the following error if I remove `externals`:

```
drake/drake/multibody/shapes/geometry.cc:2:43: fatal error: tinyobjloader/tiny_obj_loader.h: No such file or directory
```

Maybe I need to change the `include_directories` in CMakeLists.txt. Checking...
</blockquote></details>

OK I'll try to check as well. I just found it odd that none of the other externals required it. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
There appears to be a linker error:

```
/usr/bin/ld: cannot find -ltinyobjloader
collect2: error: ld returned 1 exit status
```

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-KYEj6GtBT_3AGI9T4W2:bl8zhe8)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Looks like there's an include error:

```
/home/ubuntu/workspace/linux-gcc-experimental-ros/src/drake/drake/multibody/shapes/geometry.cc:2:29: fatal error: tiny_obj_loader.h: No such file or directory
```

nkoenig commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KZ2bCSgUzaK1x74VGif:b-1jyrju) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

OK I'll try to check as well. I just found it odd that none of the other externals required it. 
</blockquote></details>

I could use a little help with this. I can make this change to the `CMakeLists.txt` file:

```
include_directories(${PROJECT_BINARY_DIR}/../install/include)
```

which is ugly. Or I can reference the `externals` directory in the source file, which is also ugly.

Is there a special cmake variable that I can use to point to the externals install include path? Or is there another recommended approach to take when using externals?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
I suspect @david-german-tri or @jamiesnape or @mwoehlke-kitware could assist with the correct CMake spelling to get that include path fixed (per the prior comment).

david-german-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KZDUvmn7iKOBvno0CKw:bt3iqqk) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary><blockquote>

I could use a little help with this. I can make this change to the `CMakeLists.txt` file:

```
include_directories(${PROJECT_BINARY_DIR}/../install/include)
```

which is ugly. Or I can reference the `externals` directory in the source file, which is also ugly.

Is there a special cmake variable that I can use to point to the externals install include path? Or is there another recommended approach to take when using externals?

</blockquote></details>

Looks like you're missing a `find_package` call in `cmake/packages.cmake`.  `tinyobjloader` already has a CMake build, so I believe find_package will Just Work.  If not, we'll have to ask @mwoehlke-kitware or @jamiesnape.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 3 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZMQKhCOmYfJGC6dbXi:-KZMQKhCOmYfJGC6dbXj:b-psfaww) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L3)):*
> ```CMake
> add_library_with_exports(LIB_NAME drakeShapes SOURCE_FILES geometry.cc element.cc visual_element.cc)
> target_link_libraries(drakeShapes spruce tinyobjloader)
> include_directories(SYSTEM ${CMAKE_SOURCE_DIR}/externals/tinyobjloader)
> ```

Do not hard code. Also always use target_include_directories.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

mwoehlke-kitware commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 3 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZMQKhCOmYfJGC6dbXi:-KZMheOl5AVHgfjiv5eU:bz9q26w) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L3)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Do not hard code. Also always use target_include_directories.
</blockquote></details>

If `target_link_libraries(drakeShapes spruce tinyobjloader)` is not sufficient (i.e. tinyobjloader does not already have interface include directories), can we fix that instead? That would be the best solution...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 3 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZMQKhCOmYfJGC6dbXi:-KZMic52bmcxPLcc9Jdk:bhcia52) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L3)):*
<details><summary><i>Previously, mwoehlke-kitware (Matthew Woehlke) wrote…</i></summary><blockquote>

If `target_link_libraries(drakeShapes spruce tinyobjloader)` is not sufficient (i.e. tinyobjloader does not already have interface include directories), can we fix that instead? That would be the best solution...
</blockquote></details>

It won't be. I am upstreaming a fix now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/CMakeLists.txt, line 3 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZMQKhCOmYfJGC6dbXi:-KZS6JyOhn8CAyXzz2LV:b-8qmi5q) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L3)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

It won't be. I am upstreaming a fix now.
</blockquote></details>

Upstream is fixed. Update the submodule SHA to `4bd75baaae2a65aba0791fac2815d6730f6127e1`.

Remove the `include_directories` line. The `target_link_libraries` call will add the correct header path.

---

*[drake/multibody/shapes/CMakeLists.txt, line 17 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZS6bcU9KSNwJ6Wzaw-:-KZS6bcU9KSNwJ6Wzaw0:be33enx) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L17)):*
> ```CMake
>     -lspruce
>     -ltinyobjloader
>   REQUIRES eigen3)
> ```

Remove `-ltinyobjloader` form `LIBS` and add `tinyobjloader` to `REQUIRES`.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-KZS708Jk1jLfE8ShR5N:bchknfk) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, david-german-tri (David German) wrote…</i></summary><blockquote>

Looks like you're missing a `find_package` call in `cmake/packages.cmake`.  `tinyobjloader` already has a CMake build, so I believe find_package will Just Work.  If not, we'll have to ask @mwoehlke-kitware or @jamiesnape.
</blockquote></details>

Add `drake_find_package(tinyobjloader CONFIG)` to `cmake/packages.cmake`. If it is always required, which it appears to be, add `REQUIRED` too.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-gcc-bazel-experimental please
@drake-jenkins-bot linux-xenial-gcc-bazel-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-xenial-clang-experimental please

jamiesnape commented:
Review status: 3 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[CMakeLists.txt, line 252 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-K_jhZVXb6dqP5B0kJsz:-K_jhZVXb6dqP5B0kJt-:bqkc6gs) ([raw file](https://github.com/robotlocomotion/drake/blob/2122c919a945bc7006240b15f8aca6a8b6afd8d6/CMakeLists.txt#L252)):*
> ```CMake
>     -DWITH_YALMIP:BOOL=${WITH_YALMIP}
>     -DWITH_YAML_CPP:BOOL=${WITH_YAML_CPP}
>     -DWITH_TINYOBJLOADER:BOOL=${WITH_TINYOBJLOADER}
> ```

Alphabetize

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-K_ji-sZdHra-b8GV2zZ:bbp2xve) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Add `drake_find_package(tinyobjloader CONFIG)` to `cmake/packages.cmake`. If it is always required, which it appears to be, add `REQUIRED` too.
</blockquote></details>

`<tinyobjloader/tiny_obj_loader.h>`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 3 of 8 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[.gitmodules, line 104 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-K_jiaPOefWJ0iVUublp:-K_jiaPOefWJ0iVUublq:b-hrdccv) ([raw file](https://github.com/robotlocomotion/drake/blob/2122c919a945bc7006240b15f8aca6a8b6afd8d6/.gitmodules#L104)):*
> ```
> [submodule "externals/tinyobjloader"]
> 	path = externals/tinyobjloader
> 	url = https://github.com/syoyo/tinyobjloader
> ```

https://github.com/syoyo/tinyobjloader.git

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 3 of 8 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[drake/lcm/test/drake_mock_lcm_test.cc, line 33 at r9](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-K_jiSUb0tGRD51x5Ixo:-K_jiSUb0tGRD51x5Ixp:b-epj7cu) ([raw file](https://github.com/robotlocomotion/drake/blob/2122c919a945bc7006240b15f8aca6a8b6afd8d6/drake/lcm/test/drake_mock_lcm_test.cc#L33)):*
> ```cpp
>       dut.get_last_published_message(channel_name);
> 
>   EXPECT_EQ(static_cast<size_t>(message_size), published_message_bytes.size());
> ```

Note: I made this change to fix a compiler artifact.  I can remove this (or change it if there is a better solution). I'm running cmake + make on xenial.

---

*[drake/multibody/shapes/CMakeLists.txt, line 3 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZMQKhCOmYfJGC6dbXi:-K_jiMqHIKNRlpbi8PTn:b-x3t9xn) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L3)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Upstream is fixed. Update the submodule SHA to `4bd75baaae2a65aba0791fac2815d6730f6127e1`.

Remove the `include_directories` line. The `target_link_libraries` call will add the correct header path.
</blockquote></details>

Thanks for the help. Changes have been made.

---

*[drake/multibody/shapes/CMakeLists.txt, line 17 at r8](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KZS6bcU9KSNwJ6Wzaw-:-K_jiGrY7pDueEMUrwxh:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/134d495ce14b3757c42f86cc1dc3d458b3756a8a/drake/multibody/shapes/CMakeLists.txt#L17)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Remove `-ltinyobjloader` form `LIBS` and add `tinyobjloader` to `REQUIRES`.
</blockquote></details>

Done.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-K_jiIkOdhxN_R39rgwR:bv19dfo) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

Add `drake_find_package(tinyobjloader CONFIG)` to `cmake/packages.cmake`. If it is always required, which it appears to be, add `REQUIRED` too.
</blockquote></details>

Thanks. I added a required find package line.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-K_jj9ks5nSGWZrsEela:b-efm4mj) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, nkoenig (Nate Koenig) wrote…</i></summary><blockquote>

Thanks. I added a required find package line.
</blockquote></details>

`<tiny_obj_loader.h>`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
Review status: 2 of 8 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[drake/multibody/shapes/geometry.cc, line 2 at r6](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-KXr2bGXJVKmyBiAa_YG:-K_jjQp2WjAiv2j4ej2I:bpfugyi) ([raw file](https://github.com/robotlocomotion/drake/blob/918372b4ebc7ea5cbe0ba3db9b5cdca490da8364/drake/multibody/shapes/geometry.cc#L2)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

`<tiny_obj_loader.h>`
</blockquote></details>

Right, thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Please rebase on master.

---

Reviewed 1 of 7 files at r6, 4 of 5 files at r9, 3 of 3 files at r10.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-K_jkCEI-CQAmkiszhAX:b-kv7t0q)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Try adding these files to fix the Bazel jobs:

WORKSPACE:
```
new_git_repository(
    name = "tinyobjloader",
    remote = "https://github.com/syoyo/tinyobjloader.git",
    commit = "7c3206f9193ca259b73fec6556cdcab45c1f5723",
    build_file = "tools/tinyobjloader.BUILD",
)
```

tools/tinyloader.BUILD:
```
# -*- python -*-

cc_library(
    name = "tinyobjloader",
    srcs = [
        "tiny_obj_loader.cc",
    ],
    hdrs = [
        "tiny_obj_loader.h",
    ],
    includes = ["."],
    linkstatic = 1,
    visibility = ["//visibility:public"],
)

liangfok commented:
Reviewed 3 of 5 files at r9, 2 of 3 files at r10, 6 of 6 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[drake/automotive/test/automotive_simulator_test.cc, line 77 at r11](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-K_lkXyBjCgGOZF1tvp0:-K_lkXyBjCgGOZF1tvp1:b-qw1kce) ([raw file](https://github.com/robotlocomotion/drake/blob/20f5632ede6fd5ddcc4fbeca69fc1fffe78c51b0/drake/automotive/test/automotive_simulator_test.cc#L77)):*
> ```cpp
>   const std::vector<const RigidBody<double>*> vehicle_bodies =
>       tree.FindModelInstanceBodies(model_instance_id);
>   EXPECT_EQ(vehicle_bodies.size(), static_cast<size_t>(num_vehicle_bodies));
> ```

This should be `static_cast<int>(vehicle_bodies.size())` rather than the other way around per style guide:

https://google.github.io/styleguide/cppguide.html#Integer_Types

(Same comment below.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Reviewed 1 of 3 files at r10.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
Add `"@tinyobjloader//: tinyobjloader",` after this line:

https://github.com/RobotLocomotion/drake/blob/master/drake/multibody/shapes/BUILD#L24

ggould-tri commented:
Reviewed 7 of 12 files at r13.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 6 files at r11, 8 of 12 files at r13.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805#-:-Ka9UxYhsoLj6T5pWmgZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 4 files at r14.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3805)*
<!-- Sent from Reviewable.io -->

nkoenig commented:
I need to fix the commits, and then I'll resubmit.
